// office-ui-fabric-react/lib/components/Link/Link.doc.js
import * as React from 'react';
import { LinkBasicExample } from './examples/Link.Basic.Example';
var LinkBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Link/examples/Link.Basic.Example.tsx');
export var LinkPageProps = {
    title: 'Link',
    componentName: 'Link',
    componentUrl: 'https://github.com/microsoft/fluentui/tree/master/packages/office-ui-fabric-react/src/components/Link',
    examples: [
        {
            title: 'Link',
            code: LinkBasicExampleCode,
            view: React.createElement(LinkBasicExample, null),
            styles: function (_a) {
                var theme = _a.theme;
                // UHF overrides. :( These are here rather than in the example because they're not necessary
                // under normal circumstances, and including them in the example makes it more confusing.
                return {
                    root: {
                        selectors: {
                            '.ms-Link': {
                                margin: 0,
                                padding: 0,
                                overflow: 'inherit',
                                textOverflow: 'inherit',
                            },
                        },
                    },
                };
            },
        },
    ],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/Link/docs/LinkOverview.md'),
    bestPractices: require('!raw-loader!office-ui-fabric-react/src/components/Link/docs/LinkBestPractices.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true,
    allowNativeProps: true,
    nativePropsElement: ['a', 'button'],
};
